### PR TITLE
Double buffer for sequencer catch

### LINKED ISSUE #158: Implement the segment caching feature of Sequencer API
**What would you like to be added**:
Implement the segment caching feature of Sequencer API
You can refer to Leaf to do double buffer optimization
The function to be implemented is at https://github.com/mosn/layotto/blob/main/pkg/runtime/sequencer/cache.go

The design doc(in chinese) is at https://mosn.io/layotto/#/zh/design/sequencer/design

**Why is this needed**:

### PR DESCRIPTION
**Pull request by ZLBer**
**What this PR does**:

Using  double buffer to  solve the problem of sequencer local cache.

**Which issue(s) this PR fixes**:

Fixes #158 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:87]: buffer may be nil

Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:87]: d.BackUpBuffer = buffer have data race

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:138]: concurrent GetNextIdFromCache may got not inited DoubleBuffer?

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:87]: `d.InUseBuffer.to-d.InUseBuffer.from == defaultLimit` make sure only one can enter this code

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:138]: @ZLBer  yes, only return error now

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:87]: @MoonShining  yes, I don't know what to do when an asynchronous error occurs？maybe try again？

Comment (MoonShining) [reply pkg/runtime/sequencer/cache.go:87]: the assignment is not guaranteed to be seen in other goroutine

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [reply pkg/runtime/sequencer/cache.go:138]: then getId will fail, which should be prevented

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [reply pkg/runtime/sequencer/cache.go:87]: how about check BackUpBuffer exist before swap

Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:87]: @MoonShining  oh ,i see , you mean this assignment must under lock?

Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:87]: @MoonShining  ok, check buffer is nil, we get it again , if still nil  then return error

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:138]: @MoonShining  I'm not very familiar with the features of golang. maybe we can  block other goroutine with  `sync.WaitGroup` ?

Comment (MoonShining) [reply pkg/runtime/sequencer/cache.go:138]: use rwmutex and a map is simpler

**Comment by seeflood** [comment]
@ZLBer Hi,the CI failed and u should use `go mod tidy` to reformat go.mod

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:147]: access store every time is better to be prevented

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:158]: use RLock here

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:83]: the behaviour we expedct is:  if `getId` execeed some threshold, `getNewBuffer` asynchoronus, which won't impact `getId` keep going. but with the huge lock here, async getNewBuffer will **block** getId.

the `DoubleBuffer` is a classic design, but implement it have challenge.

how about do more research and post what you found in this PR.

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:147]: @MoonShining  `size=0` will only return true/false  , will not request components, such as redis

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:158]: @MoonShining  is it necessary ？why

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:83]: @MoonShining  yeah,  it‘s  a problem. java has `volatile` for visibility.  I don't know what to do on golang. I will do more  research. thanks for your review.

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [reply pkg/runtime/sequencer/cache.go:147]: so the behaviour rely on the internal implementation in specific store. other stores may not care about size=0.

is it better to expose some ability check interface on store? eg: `store.Support("SegmentGet")`

Comment (seeflood) [reply pkg/runtime/sequencer/cache.go:147]: @MoonShining It's a long story...The first version of the component interface definition has a support() method, but after discussion I removed it ,see https://github.com/mosn/layotto/issues/158#issuecomment-890322241
So actually this invocation here just means "check whether this component support it"
If u guys also think it is weird,I will submit a PR to add the 'support' method back. But I think it's ok here in this PR

File: pkg/runtime/sequencer/cache.go
Comment (seeflood) [inline pkg/runtime/sequencer/cache.go:160]: I find a new problem here: BufferCatch's key is just the requested key without component name,but unfortunately different components might have same key, so they might affect each other   :3

**Review by seeflood — CHANGES_REQUESTED** [review]

**Comment by seeflood** [comment]
I added some comments in your branch when reading the code  :>
It seems challenging to implement such a concurrent safe cache with shared memory programming style in golang. I will do some research on golang concurrent programming.
Anyway,it's an interesting feature to learn and practice concurrent programming in golang,since golang told u "do not communicate by sharing memory",haha

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:160]: @seeflood  can one layotto used by  more than one application ?

**Comment by ZLBer** [comment]
@seeflood  Yes it is, simple problems become complicated.  maybe  we can simplify the problem, we first only use one buffer to implement it.

File: pkg/runtime/sequencer/cache.go
Comment (seeflood) [reply pkg/runtime/sequencer/cache.go:160]: @ZLBer Currently no, but maybe yes in the future.
But even if layotto serve only one application,this app can use two sequencers with two different `storeName`,and it will lead to conflicts here

**Comment by seeflood** [comment]
@ZLBer Actually I'm trying to write a universal DoubleBuffer in golang (for fun)
,and maybe u can search for any existing lib providing DoubleBuffer.
It's better to use a double buffer here,but if there isn't any existing lib,I think it's OK to simplify the problem now and leave the optimization work later in the future ( in some other iteration)

**Comment by seeflood** [comment]
I just found it was closed by bot. 
@ZLBer are you still working on it? 
Or I can assign it to others if you want to focus on other issues(e.g. integrating with Skywalking)

**Comment by ZLBer** [comment]
@seeflood  Any good idea to implement this ？

**Comment by ZLBer** [comment]
@seeflood @MoonShining  I use `atomic.StorePointer ` `atomic.LoadPointer` to achieve visibility to avoid locking， but  test performance is much worse than locking.  Any good idea ？

**Comment by seeflood** [comment]
> @seeflood @MoonShining I use `atomic.StorePointer ` `atomic.LoadPointer` to achieve visibility to avoid locking， but test performance is much worse than locking. Any good idea ？

!! That sounds weird
You can push your code into your forked repo and put your performance test result here,so that we can take a look at what happened

**Comment by ZLBer** [comment]
@seeflood  I changed it again, the performance has improved, a little bit worse than locking.
look at this branch : https://github.com/ZLBer/layotto/tree/id_catch_cas

**Comment by seeflood** [comment]
@zhenjunMa and @MoonShining will help review your branch and try to find the reason;
Let me review the consul lock PR

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:160]: will panic， use RLock to protect map access

**Comment by MoonShining** [comment]
hi @ZLBer，still several problem must be solved

1.  concurrent map access must be protect by RLock or Lock
2. async `getNewBuffer` should't block `getId`


how about use `CSP` model? the idea is simple


```
// store next buffers
bufChan := make(chan Buffer, 1)

// use async goroutine to getNewBuffer
go func(){
  for {
    buf := getNewBuffer()
    bufChan <- buf
  }
}()

// and in swap, just get next buffer from bufChan
```

**Comment by ZLBer** [comment]
@MoonShining  change the  backupBuffer type  to `channel` , add  a cas status `processing` means getting new buffer. the performance is similar to locking.  thanks for your review again~

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:236]: should prevent do map read twice

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:141]: should check BackUpBufferChan

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:107]: how about retry until `getNewBuffer` success?
so in swap, you can just wait on BackUpBufferChan or timeout.

Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:107]: @MoonShining   Gooood idea, put retry here and can  remove processing status

File: pkg/runtime/sequencer/cache.go
Comment (MoonShining) [inline pkg/runtime/sequencer/cache.go:99]: shoudl loop forever until success

otherwise, once defaultRetry exceeded, getId() alwasy return err

Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go:99]: when  error occurs,  need to wait some time before trying again ?

**Review by MoonShining — APPROVED** [review]
LGTM

File: pkg/runtime/sequencer/cache.go
Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go]: @seeflood  see this question again,  an easy way to solve this is add componentName before key in  `runtime_sequencer.GetModifiedKey` method.

Comment (seeflood) [reply pkg/runtime/sequencer/cache.go]: @ZLBer You are right !
We can modify keys there like`GetModifiedLockKey` in lock_config.go or `GetModifiedStateKey` in compatibility.go
Would u like to do it? Or let me submit a new PR to fix it after this PR has been merged

Comment (ZLBer) [reply pkg/runtime/sequencer/cache.go]: @seeflood  i will fix this

Comment (seeflood) [reply pkg/runtime/sequencer/cache.go]: @ZLBer Cool
You can make it support different naming strategy ,like `GetModifiedLockKey` in lock_config.go

**Comment by ZLBer** [comment]
@seeflood  add seq key prefix，just move the lock_config method  to sequencer

**Comment by seeflood** [comment]
@ZLBer Thanks ,I will read through it this noon, when eating my lunch.

By the way , please use `go fmt` to fix the CI error

File: pkg/runtime/runtime.go
Comment (seeflood) [inline pkg/runtime/runtime.go:435]: not "save lock configuration",it's sequencer  :)

**Review by seeflood — CHANGES_REQUESTED** [review]

**Comment by ZLBer** [comment]
> @ZLBer Thanks ,I will read through it this noon, when eating my lunch.
> 
> By the way , please use `go fmt` to fix the CI error 

i do the `go fmt`, but still hava  ci error,  Is there something wrong with the way I do it？

**Comment by seeflood** [comment]
@ZLBer Yep,it seems that we should run `go mod tidy`.
I did it in your branch. Let's wait for the CI

**Comment by seeflood** [comment]
@ZLBer now the ut failed :0
You can fix it according to the CI error messages

**Comment by ZLBer** [comment]
@seeflood   might too many  goroutine in ut, remove a ut..

**Comment by seeflood** [comment]
@ZLBer the CI failed again because the test coverage went down :(
Could u please add some other ut (just to keep the test coverage)

**Comment by ZLBer** [comment]
@seeflood  all is fine,  thanks

**Review by MoonShining — APPROVED** [review]

**Review by seeflood — APPROVED** [review]

**Comment by seeflood** [comment]
Thanks again for your contribution!
By the way,your new avatar is interesting @ZLBer

**Comment by ZLBer** [comment]
@MoonShining @seeflood  hahhh, thanks for your review, learned a lot.

**Pull request merged**
